dali/image/generic_image.cc

```diff
@@ -22,21 +22,38 @@ GenericImage::GenericImage(const uint8_t *encoded_buffer, size_t length, DALIIma
         Image(encoded_buffer, length, image_type) {
 }
 
-
 std::pair<std::shared_ptr<uint8_t>, Image::Shape>
 GenericImage::DecodeImpl(DALIImageType image_type,
                          const uint8_t *encoded_buffer,
                          size_t length) const {
+  auto shape = PeekShapeImpl(encoded_buffer, length);
+  int C = NumberOfChannels(image_type, shape[2]);
+  int flags = 0;
+  if (image_type == DALI_ANY_DATA && C <= 3) {
+    image_type = C == 3 ? DALI_RGB : DALI_GRAY;
+  }
+  if (image_type == DALI_ANY_DATA) {
+    flags = cv::IMREAD_UNCHANGED;
+  } else if (image_type == DALI_GRAY) {
+    flags = cv::IMREAD_GRAYSCALE;
+  } else {
+    flags = cv::IMREAD_COLOR;
+  }
+
   // Decode image to tmp cv::Mat
   cv::Mat decoded_image = cv::imdecode(
-    cv::Mat(1, length, CV_8UC1, (void *) (encoded_buffer)),         //NOLINT
-    IsColor(image_type) ? cv::IMREAD_COLOR : cv::IMREAD_GRAYSCALE | cv::IMREAD_IGNORE_ORIENTATION);
+    cv::Mat(1, length, CV_8UC1, const_cast<unsigned char*>(encoded_buffer)), flags);
```

Comment:
```suggestion
    cv::Mat(1, length, CV_8UC1, const_cast<unsigned char*>(encoded_buffer)), flags | cv::IMREAD_IGNORE_ORIENTATION);
```
We do not support EXIF yet. Please don't remove the `cv::IMREAD_IGNORE_ORIENTATION` flag.

Reply:
:+1:

Reply:
Good catch

Reply:
Actually, the flat should probably go to lines 38 and 40 - otherwise it would be likely that when adding EXIF support we'd simply remove it here and the images loaded with ANY_DATA would have their orientation removed.